### PR TITLE
Cn5zero vectors

### PR DESCRIPTION
**Pull request by luminoso-beaudoin**
For reference, here are evaluation results from running this code:

    cn5zero-vectors) rbeaudoin@secretariat:~/projects/cn5zero-vectors/conceptnet5$ cn5-vectors 
    evaluate -l data/vectors/numberbatch.h5
                                            acc           low           high
    men3000                0.864430  0.852914  0.875104
    rw                           0.633740  0.600773  0.664554
    mturk                      0.824830  0.800864  0.846158
    simlex                     0.640676  0.602597  0.675845
    gur350-de               0.786326  0.742717  0.823292
    zg222-de                0.516461  0.412863  0.606867
    ws353                     0.816783  0.778798  0.848799
    ws353-es                0.708732  0.652543  0.757169
    ws353-ro                 0.676448  0.615433  0.729404
    pku500-zh               0.523280  0.456537  0.584166
    tmu-rw-ja                 0.390513  0.365261  0.415190
    semeval-2a-en        0.782978  0.746530  0.814739
    semeval-2a-de        0.752512  0.711787  0.788193
    semeval-2a-es        0.765945  0.727078  0.799916
    semeval-2a-it          0.736363  0.693436  0.774079
    semeval-2a-fa         0.519949  0.452838  0.581185
    semeval-2b-en-de   0.779396  0.752577  0.803632
    semeval-2b-en-es   0.777944  0.751917  0.801549
    semeval-2b-en-it     0.774132  0.747609  0.798189
    semeval-2b-en-fa    0.616825  0.575854  0.654705
    semeval-2b-de-es   0.760084  0.731947  0.785633
    semeval-2b-de-it     0.761761  0.733090  0.787729
    semeval-2b-de-fa    0.611939  0.569057  0.651501
    semeval-2b-es-it     0.759374  0.731335  0.784850
    semeval-2b-es-fa    0.629007  0.588150  0.666661
    semeval-2b-it-fa      0.606681  0.564072  0.646067
                                            bias           low          high
    beliefs                     0.125565  0.092409  0.158721
    ethnicity-coarse      0.076057  0.061895  0.090220
    ethnicity-fine           0.117802  0.104171  0.131432
    ethnicity-names      0.040464  0.030850  0.050077
    gender                    0.014529  0.006930  0.022127

### CODE REVIEW

**Comment by rspeer** [comment]
Hm. The thing I was thinking of that would actually improve the situation in non-English (and allow us to try other kinds of OOV) was to use a larger vocabulary for post-retrofitting propagation (the part you call "cleanup") than for retrofitting.

For example, a German word connected to one English word should get the vector of that English word, although including it in the vocabulary used for retrofitting would be unhelpful and wasteful.

Or did you try this approach already?

**Comment by luminoso-beaudoin** [comment]
Latest evaluation results (with propagation of vectors to the full ConceptNet graph):

```

(cn5zero-vectors) rbeaudoin@secretariat:~/projects/cn5zero-vectors/conceptnet5$ cn5-vectors evaluate -l data/vectors/numberbatch.h5 
                       acc       low      high
men3000           0.865255  0.853804  0.875869
rw                0.637724  0.605030  0.668272
mturk             0.825491  0.801606  0.846746
simlex            0.641114  0.603070  0.676248
gur350-de         0.771162  0.724936  0.810473
zg222-de          0.508104  0.403420  0.599660
ws353             0.815468  0.777244  0.847695
ws353-es          0.705787  0.649146  0.754643
ws353-ro          0.703079  0.646026  0.752320
pku500-zh         0.494013  0.424720  0.557552
tmu-rw-ja         0.368954  0.343231  0.394124
semeval-2a-en     0.765623  0.726715  0.799634
semeval-2a-de     0.756183  0.715966  0.791398
semeval-2a-es     0.763451  0.724235  0.797743
semeval-2a-it     0.744911  0.703144  0.781555
semeval-2a-fa     0.599925  0.540668  0.653256
semeval-2b-en-de  0.776847  0.749764  0.801330
semeval-2b-en-es  0.770654  0.743898  0.794943
semeval-2b-en-it  0.766854  0.739603  0.791594
semeval-2b-en-fa  0.666472  0.629594  0.700355
semeval-2b-de-es  0.766097  0.738557  0.791083
semeval-2b-de-it  0.771415  0.743734  0.796454
semeval-2b-de-fa  0.663824  0.625345  0.699083
semeval-2b-es-it  0.764340  0.736792  0.789353
semeval-2b-es-fa  0.681458  0.645115  0.714723
semeval-2b-it-fa  0.658739  0.620436  0.693906
                      bias       low      high
beliefs           0.123566  0.089587  0.157544
ethnicity-coarse  0.073614  0.059069  0.088159
ethnicity-fine    0.120652  0.107220  0.134085
ethnicity-names   0.040185  0.030989  0.049382
gender            0.014320  0.007080  0.021561

```

**Comment by rspeer** [comment]
It looks like we're getting good non-English vectors out of this process. But why did English SemEval take such a dive? We've been getting over .79 for several versions.

Out of curiosity, is the number notably higher for numberbatch-biased.h5 instead of numberbatch.h5? I think I've been inconsistent about which one I evaluate, in this state where debiasing requires too much memory, and I'm concerned it might be making a big difference in English.

If it's not that, then it means that propagation is having a negative effect on English.

**Comment by luminoso-beaudoin** [comment]
After removing English terms from the propagated vocabulary, English SemEval gets better, though still not quite up to ConceptNet5.6rc, and a few other tests are slightly worse:

```

(cn5zero-vectors) rbeaudoin@secretariat:~/projects/cn5zero-vectors/conceptnet5$ cn5-vectors evaluate -l data/vectors/numberbatch.h5 
                       acc       low      high
men3000           0.864903  0.853424  0.875542
rw                0.635044  0.602166  0.665772
mturk             0.824753  0.800778  0.846090
simlex            0.641652  0.603651  0.676743
gur350-de         0.771186  0.724963  0.810493
zg222-de          0.509862  0.405404  0.601177
ws353             0.817986  0.780221  0.849810
ws353-es          0.705883  0.649257  0.754726
ws353-ro          0.704656  0.647843  0.753674
pku500-zh         0.491359  0.421844  0.555132
tmu-rw-ja         0.369876  0.344173  0.395025
semeval-2a-en     0.782925  0.746469  0.814692
semeval-2a-de     0.756879  0.716759  0.792005
semeval-2a-es     0.762079  0.722671  0.796546
semeval-2a-it     0.744145  0.702274  0.780885
semeval-2a-fa     0.598634  0.539241  0.652100
semeval-2b-en-de  0.781936  0.755385  0.805923
semeval-2b-en-es  0.777723  0.751673  0.801349
semeval-2b-en-it  0.774032  0.747498  0.798099
semeval-2b-en-fa  0.672455  0.636092  0.705838
semeval-2b-de-es  0.764774  0.737102  0.789884
semeval-2b-de-it  0.770014  0.742189  0.795188
semeval-2b-de-fa  0.662942  0.624384  0.698276
semeval-2b-es-it  0.763149  0.735483  0.788273
semeval-2b-es-fa  0.680472  0.644041  0.713824
semeval-2b-it-fa  0.657516  0.619109  0.692786
                      bias       low      high
beliefs           0.123584  0.091995  0.155173
ethnicity-coarse  0.074882  0.060479  0.089284
ethnicity-fine    0.118617  0.105282  0.131953
ethnicity-names   0.040144  0.030864  0.049425
gender            0.014518  0.007105  0.021931

```
I'm looking into the discrepancy on semeval-en.

File: conceptnet5/builders/reduce_assoc.py [hunk 1]
```diff
@@ -21,11 +23,91 @@ def concept_is_bad(uri):
             uri.startswith('/a/') or uri.count('/') <= 2)
 
 
-def reduce_assoc(filename, output_filename, cutoff=3, en_cutoff=3):
-    """
-    Takes in a file of tab-separated simple associations, and removes
-    uncommon associations and associations unlikely to be useful.
+class Graph:
```
Comment (rspeer) [inline conceptnet5/builders/reduce_assoc.py]: I'd say this class name is overly broad.

**Comment by rspeer** [comment]
Aside from my quibble about the class name, I think this is good and ready to merge.

File: conceptnet5/builders/reduce_assoc.py (hunk 1 above)
Comment (luminoso-beaudoin) [reply conceptnet5/builders/reduce_assoc.py]: It started out as something to be used in one file only; now that it is referenced elsewhere I'd have to agree; I'll change it.

**Pull request merged**